squidpy/pl/_spatial.py

```diff
@@ -0,0 +1,764 @@
+            )
+
+
+def _get_unique_map(dic: Mapping[str, Any]) -> Any:
+    """Get intersection of dict values."""
+    return sorted(set.intersection(*map(set, dic.values())))
+
+
+def _maybe_get_list(var: Any, _type: Any, ref: Sequence[Any] | None = None) -> Sequence[Any] | Any:
+    if isinstance(var, _type):
+        if ref is None:
+            return [var]
+        else:
+            return [var for _ in ref]
+    else:
+        if isinstance(var, list):
+            if (ref is not None) and (len(ref) != len(var)):
+                raise ValueError(f"Var len: {len(var)} is not equal to ref len: {len(ref)}. Please Check.")
+            else:
+                return var
+        else:
+            raise ValueError(f"Can't make list from var: `{var}`")
+
+
+def _get_seq_vminmax(
+    vmin: Any, vmax: Any, vcenter: Any, norm: Any
+) -> Tuple[Sequence[Any], Sequence[Any], Sequence[Any], Sequence[Any]]:
+    if isinstance(vmax, str) or not isinstance(vmax, Sequence):
+        vmax = [vmax]
+    if isinstance(vmin, str) or not isinstance(vmin, Sequence):
+        vmin = [vmin]
+    if isinstance(vcenter, str) or not isinstance(vcenter, Sequence):
+        vcenter = [vcenter]
+    if isinstance(norm, Normalize) or not isinstance(norm, Sequence):
+        norm = [norm]
+
+    return vmin, vmax, vcenter, norm
+
+
+def _get_source_vec(
+    adata: AnnData,
+    value_to_plot: str | None,
+    use_raw: Optional[bool | None] = None,
+    alt_var: Optional[str | None] = None,
+    layer: Optional[str | None] = None,
+    groups: Optional[Sequence[str] | str | None] = None,
+) -> NDArrayA | pd.Series:
+
+    if value_to_plot is None:
+        return np.broadcast_to(np.nan, adata.n_obs)
+    if alt_var is not None and value_to_plot not in adata.obs.columns and value_to_plot not in adata.var_names:
+        value_to_plot = adata.var.index[adata.var[alt_var] == value_to_plot][0]
+    if use_raw and value_to_plot not in adata.obs.columns:
+        values = adata.raw.obs_vector(value_to_plot)
+    else:
+        values = adata.obs_vector(value_to_plot, layer=layer)
+    if groups and is_categorical_dtype(values):
+        values = values.replace(values.categories.difference(groups), np.nan)
+    return values
+
+
+def _get_color_vec(
+    adata: AnnData,
+    value_to_plot: str | None,
+    values: NDArrayA | pd.Series,
+    palette: Optional[Sequence[str] | str | Cycler | None] = None,
+    na_color: Optional[str | Tuple[float, ...] | None] = None,
+) -> Tuple[NDArrayA, bool]:
+    to_hex = partial(colors.to_hex, keep_alpha=True)
+    if value_to_plot is None:
+        return np.broadcast_to(to_hex(na_color), adata.n_obs), False
+    if not is_categorical_dtype(values):
+        return values, False
+    elif is_categorical_dtype(values):
+        # use scanpy _get_palette to set palette if not present
+        color_map = {k: to_hex(v) for k, v in _get_palette(adata, value_to_plot, palette).items()}
```

Comment:
this is actually a point we should consider, since indeed we are subsetting and so can't do it the scanpy way. Shall we handle it at the beginning and just add a palette color if missing the way scanpy does? thoughts @hspitzer ?

Reply:
actually, you ofc were right, changed the way we handle it and the palette if not present is set upstream